### PR TITLE
Added support for filtering on parent fields and including children queries

### PR DESCRIPTION
**Pull request by jongpie**
I've done a good bit of cleanup during this change, so there are some changes that will break existing code. Here are the items that will need to be updated

1. QueryPredicate.cls has evolved into QueryFilter.cls (QueryPredicate no longer exists). I've created a small wiki page with more details https://github.com/jongpie/NebulaFramework/wiki/Filtering-Queries
    - To update your code, these method callls..
       `.whereField(Schema.Case.CreatedById, Schema.User.IsActive, QueryOperator.EQUALS, true)`
       should be replaced with 
       `.filterBy(new QueryFilter(Schema.Case.CreatedById, Schema.User.IsActive, QueryOperator.EQUALS, true))`

2. QueryBuilder.whereField() method has been replaced by QueryBuilder.filterBy() - using QueryPredicate (now QueryFilter) for everything seemed like a much approach.
3. SObjectRepository.queryFactory is now SObjectRepository.Query

Now the good bits...
1. SObjectRepository has a new constructor that will automatically add all fields to the query for the specified SObject type - essentially, it's 'select * from sobject'. I've added more details to this wiki page https://github.com/jongpie/NebulaFramework/wiki/Using-SObject-Repositories
2. QueryFilter can handle filtering on the base object, parent object fields and grandparent fields
3. Children records can be included in a parent query. Example: query for accounts and their contacts. This only counts as 1 query, even though 2 objects are being returned. Sample method code:
    ```
        public Account getAccountAndContactsWithEmails(Id accountId) {
        return (Account)this.Query
            .filterBy(new QueryFilter(Schema.Account.Id, QueryOperator.EQUALS, accountId))
            .includeChildrenRecords(
                // The sobject field of the lookup (in this case, contact's lookup to account) is used to include children records
                Schema.Contact.AccountId,
                // SObjectRepo has a new overloaded method, filterBy
                // You can use filterBy to filter the children records that are included
                new ContactRepository().filterBy(new QueryFilter(Schema.Contact.Email, QueryOperator.NOT_EQUAL_TO, null))
            )
            .getFirstQueryResult();
    }
    ```

### CODE REVIEW

**Comment by jongpie** [comment]
Fixes #13 and fixes #14 and fixes #19

File: src/classes/SObjectTypeDescriber.cls [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        Boolean sobjectTypesMatch            = sobjectFieldDescriber.validateSObjectType(this.SObjectType);
+        Boolean sobjectTypesHasMatchingField = fieldMap.containsKey(sobjectFieldDescriber.getFieldName());
+
+        return sobjectTypesMatch && sobjectTypesHasMatchingField;
+    }
+
+    private Map<Schema.SObjectField, Schema.ChildRelationship> getSObjectTypeChildRelationshipMap(Schema.SObjectType parentSObjectType) {
+        Map<Schema.SObjectField, Schema.ChildRelationship> childRelationshipMap = new Map<Schema.SObjectField, Schema.ChildRelationship>();
+
+        for(Schema.ChildRelationship childRelationship : parentSObjectType.getDescribe().getChildRelationships()) {
```
Comment (jamessimone) [inline src/classes/SObjectTypeDescriber.cls:37]: wow. I had no clue that the ChildRelationship type existed.  Awesome

File: src/classes/ISObjectRepository.cls [hunk 2]
```diff
@@ -4,29 +4,23 @@
...
     SObject getById(Id recordId);
     List<SObject> getById(List<Id> recordIdList);
-    List<SObject> get(List<QueryPredicate> queryPredicates);
-    List<SObject> get(QueryPredicate queryPredicate);
+    List<SObject> get(IQueryFilter queryFilter);
+    List<SObject> get(List<IQueryFilter> queryFilters);
     List<SObject> getByFieldForIds(Schema.SObjectField field, Object value, List<Id> idList);
     List<SObject> getByFieldForIds(Schema.SObjectField field, Object value, Set<Id> idSet);
-    List<SObject> getCreatedSinceTimeValue(Object value);
+    List<SObject> getCreatedSince(Object dateOrDatetime);
```
Comment (jamessimone) [inline src/classes/ISObjectRepository.cls]: I think removing this method is worth talking about.  The rest of the methods aren't nearly as assumptive, and you can easily accomplish the same thing with the regular .get method.  Thoughts?

Comment (jongpie) [reply src/classes/ISObjectRepository.cls]: Yeah, I'm good with removing it - I'll make that change now

**Review by jamessimone — DISMISSED** [review]
Looks good to me!

**Pull request merged**